azavea/pfb-network-connectivity
# Feature/kak/add neighborhood country#604

## Overview

Adds:
 - country field to the `Neighborhood` model and serializers
 - cross-field validation to the `Neighborhood` creation serializer to ensure the US state field is set for the US, and not set outside the US
 - check on `AnalysisJob` creation that job is for the US, to gracefully handle until #614 done
 - filtering by country


### Demo

Two serialized neighborhoods, one in US, one not:

```json
{
    "uuid": "9be694dd-3537-4922-9dcf-74334f57d268",
    "createdAt": "2018-12-05T19:00:02.016253Z",
    "modifiedAt": "2018-12-05T19:00:02.016290Z",
    "name": "boulder",
    "label": "Boulder",
    "organization": "4ff19226-d0a8-4c09-a08f-a6a56ce0e111",
    "country": "US",
    "state_abbrev": "CO",
    "boundary_file": "https://kat-pfb-storage-us-east-1.s3.amazonaws.com/neighborhood_boundaries/root/CO/boulder.zip",
    "visibility": "public",
    "last_job": null
},
{
    "uuid": "7ec58fed-983a-4e2d-8308-55c667c0c491",
    "createdAt": "2018-12-05T19:15:09.388036Z",
    "modifiedAt": "2018-12-05T19:30:43.872267Z",
    "name": "narnia",
    "label": "Narnia",
    "organization": "4ff19226-d0a8-4c09-a08f-a6a56ce0e111",
    "country": "SC",
    "state_abbrev": "",
    "boundary_file": "https://kat-pfb-storage-us-east-1.s3.amazonaws.com/neighborhood_boundaries/root/narnia.zip",
    "visibility": "public",
    "last_job": "71f377ad-7ff7-4a22-a588-1f4f28e21ad3"
}
```

Neighborhood API creation form on the back end. Country in form defaults to US, and fields are ordered to put country and state together:

![image](https://user-images.githubusercontent.com/960264/49540031-0d360c80-f89d-11e8-81d8-1169b5ec2380.png)

Error on attempting to create a neighborhood in the US without setting the state:

![pfb_missing_state_error](https://user-images.githubusercontent.com/960264/49540121-3d7dab00-f89d-11e8-91c2-19985ccc658e.png)

Error on attempting to create a neighborhood outside the US, but with a US state set:

![pfb_neighborhood_no_state_error](https://user-images.githubusercontent.com/960264/49540140-4cfcf400-f89d-11e8-9346-453d8766e113.png)


### Notes

Should we have reason to support looking up and storing administrative regions outside the US, we will probably want to switch from using the `localflavor` and `django-countries` apps currently in use for country and state lookup, and instead use [django-cities](https://github.com/coderholic/django-cities). Setting it up appears to be non-trivial, and involves importing data from `geonames`.

The front end should be unaffected by this, as it only displays neighborhoods with completed analysis jobs, and all of those will still be in the US.


## Testing Instructions

 * Run migrations
 * State should be set to US on any existing neighborhoods by the migration
 * Try creating neighborhoods inside and outside the US
 * Neighborhoods should serialize as expected
 * An analysis job created for a non-US place should immediately go to `ERROR` status
 * Filters should work as expected, like http://localhost:9200/api/neighborhoods/?country=US

Closes #604
